core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableJoinRule.java

```diff
@@ -58,35 +58,26 @@
     final RelNode right = newInputs.get(1);
     final JoinInfo info = JoinInfo.of(left, right, join.getCondition());
     if (!info.isEqui() && join.getJoinType() != JoinRelType.INNER) {
-      // EnumerableJoinRel only supports equi-join. We can put a filter on top
+      // EnumerableHashJoin only supports equi-join. We can put a filter on top
       // if it is an inner join.
-      try {
-        return EnumerableNestedLoopJoin.create(
-            left,
-            right,
-            join.getCondition(),
-            join.getVariablesSet(),
-            join.getJoinType());
-      } catch (InvalidRelException e) {
-        EnumerableRules.LOGGER.debug(e.toString());
-        return null;
-      }
+      return EnumerableNestedLoopJoin.create(
+          left,
+          right,
+          join.getCondition(),
+          join.getVariablesSet(),
+          join.getJoinType());
     } else {
       RelNode newRel;
-      try {
-        newRel = EnumerableHashJoin.create(
-            left,
-            right,
-            info.getEquiCondition(left, right, cluster.getRexBuilder()),
-            join.getVariablesSet(),
-            join.getJoinType());
-      } catch (InvalidRelException e) {
-        EnumerableRules.LOGGER.debug(e.toString());
-        return null;
-      }
-      if (!info.isEqui()) {
+      newRel = EnumerableHashJoin.create(
+          left,
+          right,
+          info.getEquiCondition(left, right, cluster.getRexBuilder()),
+          join.getVariablesSet(),
+          join.getJoinType());
+      RexNode nonEqui = info.getRemaining(cluster.getRexBuilder());
+      if (!nonEqui.isAlwaysTrue()) {
```

Comment:
This if check is redundant.

Reply:
I don't think so.

Reply:
This check is necessary. But I prefer use` !info.isEqui()` since it is more clear?.